### PR TITLE
FIX: `USE_OPTIMIZED_CONTROLS` feature flag reduces performance in built players

### PR DESCRIPTION
**Pull request by jfreire-unity**
### Description

This PR tries to make a partial fix for https://issuetracker.unity3d.com/issues/input-system-enabling-both-optimization-options-of-input-system-reduces-performance-in-built-players, related to `USE_OPTIMIZED_CONTROLS`. 
Regarding `USE_READ_VALUE_CACHING` there are no fixes to be done with this PR.

#### Optimized control read value changes

 `InputManager.OnUpdate()` was doing an extra recursive check to evaluate if the format of the control has changed to a format that cannot be optimized. This check had the goal to help users be aware of a problem and it would also fix it for them. However, even though this check was only be done in `DEVELOPMENT_BUILD`, the gains could not be noticed when using a profiler which requires `DEVELOPMENT_BUILD` to be set.
 
This leads me to conclude that a `DEVELOPMENT_BUILD` variant is not the right place for it even though we had good intention with it.
I made the check run on `EDITOR` code only (`PlayMode` will see a performance cost anyway) and updated the documentation so that users are more aware of which cases this won't work in case they start having problems. Also, I mentioned the performance cost on `PlayMode` due to these checks.

The alternative to fix this for good would be to have a **breaking change** so that `AxisControl.invert` and other parameters could call `InputControl.ApplyParameterChanges()` to tell the control if they could be optimized or not.
Since we don't want to introduce a breaking change, we could also introduce somekind of `INPUT_SYSTEM_DEBUG` that would run for `DEVELOPMENT_BUILD` but not when the profiler is used. But this would be more time consuming and the gains would be similar in my view.

#### Control value caching

There's really no changes to this code path. I added more performance tests to show examples where there are positive and negative performance impacts. Although these tests are very specific, it does prove this performance path doesn't provide performance improvements in **every use case**. The reason why this happens is that looking for "stale" controls ([doing a binary search whenever we are writing a state change](https://github.com/Unity-Technologies/InputSystem/blob/93e5db963c4499dbeaf58679bb5da918bcdf3347/Packages/com.unity.inputsystem/InputSystem/InputManager.cs#L3690-L3701)) has some overhead that might not serve any purpose in some cases.

I believe that to have better performance from this path, we would need to replace state monitors and use the logic we have to look for "stale" controls instead. 

The documentation was updated to attempt to improve communication to users with regards to expectations of this feature flag.

### Notes

These changes can be seen in the performance tests introduced in #1905 .
 Player builds now show `USE_OPTIMIZED_CONTROLS` being more performant than in `PlayMode` where the performance cost is still visible (and was still present in development player builds as well, before this PR).

PlayMode:
image

Player build (MacOSXStandalone):
image

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Review by stefanunity — APPROVED** [review]
Looks like a very targeted step fix.
Really appreciate the changes to the docs being more explicit what these flags do or don't do.

File: Packages/com.unity.inputsystem/Documentation~/Controls.md [hunk 1]
```diff
@@ -243,7 +243,7 @@ Use [`InputControl<T>.value`](../api/UnityEngine.InputSystem.InputControl-1.html
...
+When the `'USE_READ_VALUE_CACHING'` internal feature flag is set, the Input System will switch to an optimized path for reading control values. This path efficiently marks controls as 'stale' when they have been actuated and subsequent calls to [`InputControl<T>.ReadValue`](../api/UnityEngine.InputSystem.InputControl-1.html#UnityEngine_InputSystem_InputControl_1_ReadValue) will only apply control processing when absolutely necessary. Control processing in this case can mean any hard-coded processing that might exist on the control, such as with [`AxisControl`](../api/UnityEngine.InputSystem.Controls.AxisControl.html) which has built-in inversion, normalisation, scaling etc, or any processors that have been applied to the controls' [processor stack](Processors.md#processors-on-controls). This can have a significant positive impact on performance, especially when using complex composite input actions with many composite parts, such as a movement input action that could be bound to W, A, S, and D on the keyboard, two gamepad sticks and a DPad.
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/Documentation~/Controls.md]: > This can have a significant positive impact on performance, especially when using complex composite input actions with many composite parts, such as a movement input action that could be bound to W, A, S, and D on the keyboard, two gamepad sticks and a DPad.

Regarding this sentence, maybe mentioned why? That the performance gain is coming from caching a somewhat "expensive" computation instead of doing it on the fly? And that this isn't recommended fin values are only consumed once?

Do you have any insight into the actual cost here? Is it coming from memory allocation or is it memcpy or where is the cost allocated?

File: Packages/com.unity.inputsystem/Documentation~/Controls.md [hunk 2]
```diff
@@ -256,10 +256,21 @@ If there are any non-obvious inconsistencies, 'PARANOID_READ_VALUE_CACHING_CHECK
...
+> __Please note__: This optimization has a performance impact on `PlayMode` as we do extra checks to ensure that the controls have the correct memory representation during development. Don't be alarmed if you see a performance drop in `PlayMode` when using this optimization as it's expected at this stage.
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/Documentation~/Controls.md:270]: Maybe suggest to always check actual performance with a profiler in release mode build where such checks are not active?

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 3]
```diff
@@ -952,8 +952,9 @@ internal void SetOptimizedControlDataTypeRecursively()
         // This is mainly to AxisControl fields being public and capable of changing at any time even if we were not anticipated such a usage pattern.
         // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        // Only do this check in development builds and editor in hope that it will be sufficient to catch any misuse during development.
-        [Conditional("DEVELOPMENT_BUILD"), Conditional("UNITY_EDITOR")]
+        // Only do this check in and editor in hope that it will be sufficient to catch any misuse during development.
+        // It is not done in debug builds because it has a performance cost and it will show up when profiled.
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:956]: I am slightly confused here, why would it be bad to have this in a debug build? Debug mode code shouldn't be profiled unless some kind of rough ballpark number is of interest? I think checking in editor-only might be fine, but I am confused about caring about debug build and profiling numbers. I mean many non-UNity tools shows you many red warnings when trying to profile or benchmark something in a debug build.

**Comment by Pauliusd01** [comment]
I tried running the original repro project and I don't notice a difference in the mentioned "PreUpdate.NewInputUpdate" times higher than 0.01ms there - in other words I couldn't even repro the original bug. Tried to run the test from your notes both in this PR and copied to Develop - again no difference. So either I'm doing something wrong or it could be platform specific since I used Win 10 for both projects. I think I'll refrain from reviewing pointlessly

File: Packages/com.unity.inputsystem/Documentation~/Controls.md (hunk 1 above)
Comment (jfreire-unity) [reply Packages/com.unity.inputsystem/Documentation~/Controls.md]: I didn't change this area. It was a formatting change AFAIK.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 3 above)
Comment (jfreire-unity) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:956]: You can only use the Unity Profiler with development builds. See https://docs.unity3d.com/Manual/profiler-profiling-applications.html.

**Comment by jfreire-unity** [comment]
> I tried running the original repro project and I don't notice a difference in the mentioned "PreUpdate.NewInputUpdate" times higher than 0.01ms there - in other words I couldn't even repro the original bug. Tried to run the test from your notes both in this PR and copied to Develop - again no difference. So either I'm doing something wrong or it could be platform specific since I used Win 10 for both projects. I think I'll refrain from reviewing pointlessly

Thanks. I'll try to understand your issues in Slack 👍

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 3 above)
Comment (jfreire-unity) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:956]: It seems the only way to do release builds profiling on Unity is by using native profilers. I didn't not look into those and the reported issue was found using the Unity profiler as well. The issue likely didn't exist in release builds but users could not measure them with the Unity profiler.

The degradation of performance with development builds (which are the only ones users can profile) due to using an optimization flag is also not ideal . That's why I think the best is to remove that code variant at this moment. This way at least when users measure the performance of the code using optimizations, they will see some performance improvement in player builds (or at least no performance degradation in the worst case).

File: Packages/com.unity.inputsystem/Documentation~/Controls.md (hunk 2 above)
Comment (jfreire-unity) [reply Packages/com.unity.inputsystem/Documentation~/Controls.md:270]: Resolving this comment as there is no way to profile release builds in Unity at the moment.

Comment (ekcoh) [reply Packages/com.unity.inputsystem/Documentation~/Controls.md:270]: OK, if only considering Unity Profiler then that seems true. I was thinking about profiling outside the scope of what is built into Unity as in development activity using any tool.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 3 above)
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:956]: OK, I would recommend taking a look with e.g. Superluminal to make sure we understand behavior in release build.

**Review by ritamerkl — APPROVED** [review]

**Comment by jfreire-unity** [comment]
@ekcoh @stefanunity I added some new commits with regards to caching read values documentation and performance tests. Feel free to comment and approve/request changes based on my last changes. Thanks!

File: Assets/Tests/InputSystem/CorePerformanceTests.cs [hunk 4]
```diff
@@ -766,6 +766,66 @@ void CallUpdate()
...
+
+        var value = 0f;
+        var value2d = Vector2.zero;
+        Measure.Method(() =>
+        {
+            for (var i = 0; i < 1000; ++i)
+            {
+                // Make sure state changes are different from previous state so that we mark the controls as
+                // stale.
+                InputSystem.QueueStateEvent(gamepad, new GamepadState { leftTrigger = i / 1000f, rightTrigger = i / 1000f});
```
Comment (ekcoh) [inline Assets/Tests/InputSystem/CorePerformanceTests.cs]: Wouldn't it make sense to not include the queue state event part in the benchmark and instead put this on a background thread? (applies to all the benchmarks). If you do, you can spawn that thread and then let it busy-wait on an atomic until measurement thread signals it and then let them execute. Otherwise this might be dominated by the part that is not subject for testing?

File: Assets/Tests/InputSystem/CorePerformanceTests.cs [hunk 5]
```diff
@@ -766,6 +766,66 @@ void CallUpdate()
...
+            .Run();
+
+
+        Measure.Method(() =>
+        {
+            for (var i = 0; i < 1000; ++i)
+            {
+                // Make sure state changes are different from previous state so that we mark the controls as
+                // stale.
+                InputSystem.QueueStateEvent(gamepad,
```
Comment (ekcoh) [inline Assets/Tests/InputSystem/CorePerformanceTests.cs:839]: Same comment as previous test.

**Review by ekcoh — COMMENTED** [review]
In general looks good, but I think we should move the queuing part to another thread to not colour the benchmark with work that might be a majority of the work involved. Otherwise see if there is framework support to pause/resume benchmark and handle that injection outside the profiled statement.

**Review by lyndon-unity — APPROVED** [review]
Thanks for the extra documentation additions to add some more clarity.

Moving the checks to editor only (rather than development mode on player) seem reasonable, especially for the opt in optimization.

File: Assets/Tests/InputSystem/CorePerformanceTests.cs (hunk 4 above)
Comment (jfreire-unity) [reply Assets/Tests/InputSystem/CorePerformanceTests.cs]: In theory that makes sense and it's maybe something we should do. But I think it should be out of scope of this PR as it will be more complex. 
Also, I'm not sure the impact of `InputSystem.QueueStateEvent` matters that much here. It will be the same regardless of the optimization type. Also, the goal here is to do a relative benchmark rather than a absolute one. Even if there's something "coloring" the test with some extra work, it colors all the tests in the same way.

**Review by Pauliusd01 — APPROVED** [review]
Tried testing this again, seems like the performance test results window was the problem for me showing very similar results but if I copy the results directly from the generated XML file I see the improvement vs Develop. Looks good to me

**Comment by jfreire-unity** [comment]
I updated the tests and the Controls documentation with regards to the control value caching optimization path.
I was being wrongly biased by the project-wide actions in my performance measurements.
